src/etl/tissue_enrichment/tissue_enrichment.py

```diff
@@ -0,0 +1,195 @@
+            for c in value_vars
+        )
+    )
+
+    # Add to the DataFrame and explode
+    _tmp = df.withColumn("_vars_and_vals", f.explode(_vars_and_vals))
+
+    cols = id_vars + [
+        f.col("_vars_and_vals")[x].alias(x) for x in [var_name, value_name]
+    ]
+    return _tmp.select(*cols)
+
+
+def _ndtr(x: float, mean: float, std: float) -> float:
+    """This function calculates the area under the cumulative normal distribution.
+
+    Args:
+        x (float): Observed rank of the peak.
+        mean (float): Expected rank under the null hypothesis.
+        std (float): Standard deviation of the null.
+
+    Returns:
+        float: Area under the pdf.
+    """
+    return float(0.5 + 0.5 * sps.erf((x - mean) / (std * 2**0.5)))
+
+
+def m_sd(x: VectorUDT) -> float:
+    """Feed to udf for vectorized square rooting.
+
+    Args:
+        x (VectorUDT): expected variance of peak rankings.
+
+    Returns:
+        float: Standard deviation of the mean of the n peak ranks overlapping GWAS SNPs.
+    """
+    return math.sqrt(x)
+
+
+def v_cdf(x: VectorUDT, y: VectorUDT, z: VectorUDT) -> float:
+    """Calculates the p-value.
+
+    Args:
+        x (VectorUDT): observed peak rank.
+        y (VectorUDT): expected peak rank.
+        z (VectorUDT): standard deviation under the null.
+
+    Returns:
+        float:  Enrichment p-values.
+    """
+    return float(1 - _ndtr(x, z, y))
+
+
+def cheers(peaks_wide: DataFrame, snps: DataFrame) -> DataFrame:
+    """Calculates tissue enrichment based on the CHEERS method (link here).
+
+    Args:
+        peaks_wide (DataFrame): A 4 column .bed file for consensus peaks, 4th column indicates the signal intensity within each peak.
+        snps (DataFrame): LD-expanded/Credible SNPs for the trait of interest.
+
+    Returns:
+        DataFrame: The enrichment p-value.
+    """
+    peaks_overlapping = (
+        # Only need the peak coords
+        peaks_wide.select("chr", "start", "end")
+        # Do a inner join
+        .join(
+            snps,
+            (
+                (f.col("chrom") == f.col("chr"))
+                & (f.col("pos") >= f.col("start"))
+                & (f.col("pos") <= f.col("end"))
+            ),
+            how="inner",
+        )
+        .drop("chrom")
+        .select("study_id", "chr", "start", "end")
+        .distinct()
+        .persist()
+    )
+
+    mean_sd = f.udf(m_sd, t.DoubleType())
+
+    # Store total number of peaks and total overlapping peaks
+
+    n_unique_peaks = (
+        peaks_overlapping.groupBy("study_id")
+        .count()
+        .withColumnRenamed("count", "count_peaks")
+    )
+    n_total_peaks = float(peaks_wide.count())
+    n_unique_peaks = n_unique_peaks.withColumn(
+        "mean_variance", ((float(n_total_peaks) ** 2) - 1) / (12 * f.col("count_peaks"))
+    )
+    n_unique_peaks = n_unique_peaks.withColumn(
+        "mean_sd", mean_sd("mean_variance")
+    ).drop("mean_variance")
```

Comment:
I'm not 100% sure what is happening here, but I have a feeling much of this process can be sped up by using windows.

Reply:
This is estimating the expected rank mean and variance for each "peak" in the dataset as described[ in the paper:](https://www.nature.com/articles/s41588-019-0493-9#Sec9:~:text=where%20max%20and,overlap%20at%20random.)